src/core/allow-bind.c

```diff
@@ -0,0 +1,123 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "allow-bind.h"
+#include "unit.h"
+
+#if BUILD_BPF
+/* libbpf, clang and llc compile time dependencies are satisfied */
+#include "bpf-object.h"
+#include "bpf/allow_bind/allow-bind-hexdump.h"
+#include "cgroup-bpf.h"
+#include "set.h"
+
+static int prepare_bpf_object(Unit* u, const Set *ports, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        const char *map_name = "allow_ports";
+        int map_fd, r;
+        uint8_t dummy;
+        void *port;
+
+        assert(ports);
+        assert(object);
+
+        r = bpf_object_new(allow_bind_hexdump_buffer, sizeof(allow_bind_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to create BPF object from hexdump buffer: %m");
+
+        if (set_size(ports) > 0) {
+                r = bpf_object_resize_map(obj, map_name, set_size(ports));
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Failed to resize BPF map '%s': %m", map_name);
+        }
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        map_fd = bpf_object_get_map_fd(obj, map_name);
+        if (map_fd < 0)
+                return log_unit_error_errno(u, r, "Failed to get BPF map '%s' fd: %m", map_name);
+
+        SET_FOREACH(port, ports)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, r, "Failed to update BPF map '%s", map_name);
+
+        *object = TAKE_PTR(obj);
+        return 0;
+}
+
+int allow_bind_supported(void) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        int r;
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOTSUP),
+                                "Not running with unified cgroup hierarchy, BPF is not supported");
+
+        /*
+         * Load bpf object w/o populating maps to probe whether
+         * BPF_PROG_TYPE_CGROUP_SOCK_ADDR program type is supported by kernel and
+         * if resource limits permit locking enough memory.
```

Comment:
Is there a way to distinguish between the kernel not supporting BPF and actual errors that can happen when loading? If running with legacy cgroups is ENOTSUP, IMHO this should be too

Reply:
@bluca

Please see my reply to @anitazha: https://github.com/systemd/systemd/pull/17655#discussion_r542029191

libbpf logs to kernel log via `printk`, should it be duplicated in journal logs?

Reply:
there's no need if there are already logs.